File: Easy/Reverse a linked list/reverseList.cpp
#include <bits/stdc++.h>
using namespace std;
struct Node {
  int data;
  struct Node *next;
  Node(int x) {
    data = x;
    next = NULL;
  }
};

class Solution {
 public:
  struct Node *reverseList(struct Node *head) {
    stack<int> s;
    Node *temp = head;
    while (temp != NULL) {
      s.push(temp->data);
      temp = temp->next;
    }
    temp = head;
    while (temp != NULL) {
      temp->data = s.top();
      temp = temp->next;
      s.pop();
    }

    return head;
  }
};

void printList(struct Node *head) {
  struct Node *temp = head;
  while (temp != NULL) {
    printf("%d ", temp->data);
    temp = temp->next;
  }
}

int main() {
  int T, n, l, firstdata;
  cin >> T;

  while (T--) {
    struct Node *head = NULL, *tail = NULL;

    cin >> n;

    if (n != 0) {
      cin >> firstdata;
      head = new Node(firstdata);
      tail = head;
    }

    for (int i = 1; i < n; i++) {
      cin >> l;
      tail->next = new Node(l);
      tail = tail->next;
    }

    Solution ob;
    head = ob.reverseList(head);

    printList(head);
    cout << endl;
  }
  return 0;
}

File: Easy/Reverse a linked list/ReverseList.java
import java.io.*;
import java.util.*;

class Node {

    int data;
    Node next;

    Node(int x) {
        data = x;
        next = null;
    }
}

class ReverseList {

    static void printList(Node node) {
        while (node != null) {
            System.out.print(node.data + " ");
            node = node.next;
        }
        System.out.println();
    }

    public static void main(String args[]) throws IOException {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();

        while (t > 0) {
            int n = sc.nextInt();

            Node head = new Node(sc.nextInt());
            Node tail = head;

            for (int i = 0; i < n - 1; i++) {
                tail.next = new Node(sc.nextInt());
                tail = tail.next;
            }

            Solution ob = new Solution();
            head = ob.reverseList(head);
            printList(head);
            t--;
        }
    }
}

class Solution {

    Node reverseList(Node head) {
        Node prev = null;
        Node current = head;
        Node next = head.next;

        while (current != null) {
            next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        head = prev;
        return head;
    }
}
